torch/distributed/_sharded_tensor/api.py

```diff
@@ -550,9 +559,46 @@ def sharding_spec(self) -> ShardingSpec:
         """
         return self._sharding_spec
 
+    def reshard(self, resharding_spec: ReshardingSpec) -> Union[ShardedTensor, torch.Tensor]:
+        """
+        Reshards a tensor given the resharding_spec, if output_local_tensors is
+        True in resharding_spec.
+
+        It would output the local shard on each rank instead of ShardedTensor
+        (only works for a single local shard).
+        """
+        # TODO: implement the reshard for ShardedTensor to ShardedTensor.
+        output_local_tensors = resharding_spec.output_local_tensors
+        if not output_local_tensors:
+            raise RuntimeError(
+                "Reshard not supported for ShardedTensor output yet."
+            )
+        from torch.distributed._sharded_tensor.ops._common import (
+            _result_distribute_with_col_rearrange,
+        )
+        local_shard = self.local_shards()[0].tensor
+        local_results = []
+        for i in range(local_shard.size(0)):
+            local_results.append(local_shard[i].t().contiguous())
+
+        # Distribute results to each rank with col rearrangement.
+        output = _result_distribute_with_col_rearrange(
+            local_results,
+            local_shard,
+            len(self._metadata.shards_metadata),
+            self,
+            self._process_group,
+        )
+
+        # transpose the output and return result.
+        return output.t().contiguous()
```

Comment:
We are not using the `resharding_spec` here at all to determine how to reshard?

Reply:
For aggregation to a local tensor... Probably yes..

Reply:
> For aggregation to a local tensor... Probably yes..

Not sure I followed your comment. Don't we need to use `resharding_spec` here to make this function generic? To begin with, I think supporting a handful of resharding cases which fit our use cases is fine and we throw errors in other cases. But we need to have a generic implementation here.

Reply:
I see.. Sure, will do. Thanks!

Reply:
Make the reshard a generic function.